### PR TITLE
Refactor response code outof api request

### PR DESCRIPTION
**Pull request by michael-mclawhorn**
Changes to pull response processing code out of the ApiRequest and push it into the response building.

Implements: https://github.com/yahoo/fili/issues/711

### CODE REVIEW

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/ResponseFormatType.java
Comment (cdeszaq) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/ResponseFormatType.java]: This needs some docs, since it looks like it'll be a core, foundational piece of content negotiation going forward (which is a good thing!)

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/ResponseFormatType.java
Comment (cdeszaq) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/ResponseFormatType.java]: This `accepts` pattern is great, and should serve as a fantastic base for registering additional `ResponseFormatTypes` where the 1st one "asked" wins.

I can see it needing to be extended (how, I'm not quite sure yet) to handle the "q" part of an Accept header w/ multiple content types, but that's a problem for the future and shouldn't be too painful to solve.

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/PageLinkBuilder.java
Comment (cdeszaq) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/PageLinkBuilder.java]: Would be good to get a doc on the so it's clear what it's responsibilities are.

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/PageLinkBuilder.java
Comment (cdeszaq) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/PageLinkBuilder.java]: Is this supposed to be commented out? If so, can it just be removed instead? If it can't, a comment as to _why_ would ward off future would-be code-cleanup do-gooders .

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java
Comment (cdeszaq) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java]: Missing a `@param` field.

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java
Comment (cdeszaq) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java]: Is the `toString` needed? Can it be resolved to a type in the `apiRequest` so that stringly typed things are only at the boundary?

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java
Comment (cdeszaq) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java]: Should the JavaDoc for this method be different than the one above? It looks like it's doing more...

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java
Comment (cdeszaq) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java]: Should the JavaDoc for this method be different than the one above? It looks like it's doing more...

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/JobsServlet.java
Comment (cdeszaq) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/JobsServlet.java]: Definitely should resolve this before this get's merged ;-)

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/SlicesServlet.java
Comment (cdeszaq) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/SlicesServlet.java]: Can this just be deleted?

**Review by cdeszaq — DISMISSED** [review]
Overall looks pretty good. There's some duplication around formatting with those 3 methods that I think might be able to be consolidated, but nothing major.

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/PageLinkBuilder.java
Comment (michael-mclawhorn) [reply fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/PageLinkBuilder.java]: Good call.

I think I phoned it in on this class because I don't think it's a long term parking spot for these responsibilities.  Eventually I want to make the building of response headers more injectable, but I had enough change in flight I didn't want to add a lot more.

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/PageLinkBuilder.java
Comment (michael-mclawhorn) [reply fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/PageLinkBuilder.java]: Missed this one during self review.  It'll be gone with my next commit.

**Review by archolewa — COMMENTED** [review]
A few very small things. Overall, I'm very happy with this refator. I've been wanting to clean up the pagination stuff since the day it first went in.

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/ResponseFormatType.java
Comment (cdeszaq) [reply fili-core/src/main/java/com/yahoo/bard/webservice/web/ResponseFormatType.java]: Getting out of string land and into type-land earlier instead of later is always good. +1 to getting rid of the String version.

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/HttpResponseMaker.java
Comment (QubitPi) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/HttpResponseMaker.java]: It might be better to inject `ResponseUtils` rather than creating one

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/DefaultResponseFormatType.java
Comment (QubitPi) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/DefaultResponseFormatType.java]: Are we still considering JASONAPI? I know people are only using the first 3

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/DefaultResponseFormatType.java
Comment (QubitPi) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/DefaultResponseFormatType.java]: Extra lines

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/DefaultResponseFormatType.java
Comment (QubitPi) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/DefaultResponseFormatType.java]: 2018

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/DefaultResponseFormatType.java
Comment (QubitPi) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/DefaultResponseFormatType.java]: "Default response format type"?

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/ResponseFormatType.java
Comment (QubitPi) [reply fili-core/src/main/java/com/yahoo/bard/webservice/web/ResponseFormatType.java]: Javadoc needed

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/ResponseFormatType.java
Comment (QubitPi) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/ResponseFormatType.java]: 2018 or 2017?

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/ApiRequest.java
Comment (QubitPi) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/ApiRequest.java]: Really nice 👍 I would like to remove them as well since I first saw them in this class.

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/ApiRequestImpl.java
Comment (QubitPi) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/ApiRequestImpl.java]: Why are we adding case of `""`? I assume format is null is request string doesn't put format in it

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/PageLinkBuilder.java
Comment (QubitPi) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/PageLinkBuilder.java]: 2018

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/PageLinkBuilder.java
Comment (QubitPi) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/PageLinkBuilder.java]: Extra line

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java
Comment (QubitPi) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java]: inject?

**Review by ghost — CHANGES_REQUESTED** [review]

File: CHANGELOG.md
Comment (QubitPi) [inline CHANGELOG.md]: Just curious, what are the benefits of 'accepts' model?

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java
Comment (QubitPi) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java]: Could we push `containerRequestContext.getUriInfo();` down to the only place where it is used?

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java
Comment (QubitPi) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java]: Not a blocker but I see `"text/csv; charset=utf-8"` in multiple places. Should we make it a constant somewhere?

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java
Comment (QubitPi) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java]: Extra line

**Review by ghost — CHANGES_REQUESTED** [review]

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java
Comment (QubitPi) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java]: Can be inlined with the previous line

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java
Comment (QubitPi) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java]: What if we want to return non-OK response afterwards? Should we defer the status assignment?

Actually I've seen line 146 executing the same OK assignment below

**Review by ghost — CHANGES_REQUESTED** [review]
Very good refactoring 👍 Just some minor tweaks

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/ResponseFormatType.java
Comment (michael-mclawhorn) [reply fili-core/src/main/java/com/yahoo/bard/webservice/web/ResponseFormatType.java]: This bridges us across the gap from where we are to where we want to be.
The ApiRequest may build a ResponseFormatType that has request specific state (not just an enum) but we may want enum-like configuration objects to fork code paths with.

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/ApiRequestImpl.java
Comment (michael-mclawhorn) [reply fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/ApiRequestImpl.java]: oversight.

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/DefaultResponseFormatType.java
Comment (michael-mclawhorn) [reply fili-core/src/main/java/com/yahoo/bard/webservice/web/DefaultResponseFormatType.java]: We've got code for it, so we shouldn't pull it out without deprecating it for a while.

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/ApiRequestImpl.java
Comment (michael-mclawhorn) [reply fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/ApiRequestImpl.java]: This is more of an abundance of precaution in case a container replaced a null with an "" for a parameter unspecified.  Probably unnecessary.

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java
Comment (michael-mclawhorn) [reply fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java]: I'd like to postpone doing this just because I don't know if that single method utility class is long for this world.

File: CHANGELOG.md
Comment (michael-mclawhorn) [reply CHANGELOG.md]: Enum equality is broken.  You can't compare enums with instances from other class hiearchies.  What we want is an equality-like operator but in which we can do a many to one comparison where 'CSV' can accept an instance of NamedExportReportFormat which implements ReportFormatType but also has state such as the name of the export header for that report.

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java
Comment (michael-mclawhorn) [reply fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java]: Not a bad idea, but I'd like to not do that in this PR.

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java
Comment (michael-mclawhorn) [reply fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/EndpointServlet.java]: This response builder should only be used on successes.  But we can look at that more later.

**Comment by michael-mclawhorn** [comment]
Needs rereview.

**Review by ghost — APPROVED** [review]

**Review by archolewa — APPROVED** [review]
A few typos, and a suggestion for some more discussion on how to actually
inject the ResponseFormatType into the places where it needs to be. As is, in
order to provide our own response formats, we need to override several things.

But none of it is blocking.

Also, I'm going to be dismissing @cdeszaq review. It looks like everything
he requested have been addressed, they're all minor, and we haven't heard from
him recently ;).

**Pull request merged**
